src/IssueViz.Security/Hotspots/LocalHotspotsStore.cs

```diff
@@ -0,0 +1,214 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+using SonarLint.VisualStudio.IssueVisualization.Security.IssuesStore;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.Hotspots
+{
+    /// <summary>
+    /// Methods for modification of <see cref="ILocalHotspotsStore"/>
+    /// </summary>
+    public interface ILocalHotspotsStoreUpdater
+    {
+        void AddOrUpdate(string filePath, IEnumerable<IAnalysisIssueVisualization> hotspots);
+        void RemoveForFile(string filePath);
+    }
+
+    /// <summary>
+    /// Represents the storage for locally analyzed hotspots and matching them against hotspots from the server
+    /// </summary>
+    internal interface ILocalHotspotsStore : ILocalHotspotsStoreUpdater, IIssuesStore
+    {
+        List<LocalHotspot> GetAllLocalHotspots();
+    }
+    
+    [Export(typeof(ILocalHotspotsStoreUpdater))]
+    [Export(typeof(ILocalHotspotsStore))]
+    [Export(typeof(IIssuesStore))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class LocalHotspotsStore : ILocalHotspotsStore
+    {
+        private static readonly List<IAnalysisIssueVisualization> EmptyList = new List<IAnalysisIssueVisualization>();
+
+        private readonly object lockObject = new object();
+        private readonly IHotspotMatcher hotspotMatcher;
+        private readonly IThreadHandling threadHandling;
+        private readonly IServerHotspotStore serverHotspotStore;
+        private Dictionary<string, List<LocalHotspot>> fileToHotspotsMapping =
+            new Dictionary<string, List<LocalHotspot>>();
+
+        private List<SonarQubeHotspot> serverHotspots = new List<SonarQubeHotspot>();
+        private HashSet<SonarQubeHotspot> unmatchedHotspots = new HashSet<SonarQubeHotspot>();
+
+        [ImportingConstructor]
+        public LocalHotspotsStore(IServerHotspotStore serverHotspotStore, IHotspotMatcher hotspotMatcher, IThreadHandling threadHandling)
+        {
+            this.serverHotspotStore = serverHotspotStore;
+            this.hotspotMatcher = hotspotMatcher;
+            this.threadHandling = threadHandling;
+            InitializeServerHotspots(this.serverHotspotStore.GetAll());
```

Comment:
This is problematic, based on what @bigfluffycookie pointed out this morning - the VS threading cookbook [says](https://github.com/microsoft/vs-threading/blob/main/doc/cookbook_vs.md#how-do-i-effectively-verify-that-my-code-is-fully-free-threaded) importing constructors should be free-threaded i.e. able to run completely on the caller's thread
-> we might be on the UI thread
-> we shouldn't be making server calls

Since importing constructors also have to be transitively free-thread there won't be a lot we can do. We'll need to come up with a policy on what do in importing constructors. The simplest policy would be to do nothing except register event handlers. 

For now, it might be easiest to just drop the initial refresh and revisit how we do it later.
There are a couple of options: one is just relying on the timed update to kick in at some point; another would be to move the responsibility to the package (i.e. have the package explicitly call for a refresh/trigger the update).

Reply:
Actually, what does `serverHotspotStore.GetAll()` do? Just it just fetch the issues currently known to the server hotspot store? If so it might be ok to call in in the constructor.

Reply:
Yeah it should just look up the local value and not fetch from the server